sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/reflect/DoFnSignaturesTest.java

```diff
@@ -378,6 +382,105 @@ public void foo(ProcessContext context) {}
   }
 
   @Test
+  public void testStateParameterNoAnnotation() throws Exception {
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage("missing StateId annotation");
+    thrown.expectMessage("myProcessElement");
+    thrown.expectMessage("index 1");
+    DoFnSignature sig =
+        DoFnSignatures.INSTANCE.getSignature(
+            new DoFn<KV<String, Integer>, Long>() {
+              @ProcessElement
+              public void myProcessElement(
+                  ProcessContext context, ValueState<Integer> noAnnotation) {}
+            }.getClass());
+  }
+
+  @Test
+  public void testStateParameterUndeclared() throws Exception {
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage("undeclared");
+    thrown.expectMessage("my-state-id");
+    thrown.expectMessage("myProcessElement");
+    thrown.expectMessage("index 1");
+    DoFnSignature sig =
+        DoFnSignatures.INSTANCE.getSignature(
+            new DoFn<KV<String, Integer>, Long>() {
+              @ProcessElement
+              public void myProcessElement(
+                  ProcessContext context, @StateId("my-state-id") ValueState<Integer> undeclared) {}
+            }.getClass());
+  }
+
+  @Test
+  public void testStateParameterDuplicate() throws Exception {
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage("duplicates");
+    thrown.expectMessage("my-state-id");
+    thrown.expectMessage("myProcessElement");
+    thrown.expectMessage("index 2");
+    DoFnSignature sig =
+        DoFnSignatures.INSTANCE.getSignature(
+            new DoFn<KV<String, Integer>, Long>() {
+              @StateId("my-state-id")
+              private final StateSpec<Object, ValueState<Integer>> myfield =
+                  StateSpecs.value(VarIntCoder.of());
+
+              @ProcessElement
+              public void myProcessElement(
+                  ProcessContext context,
+                  @StateId("my-state-id") ValueState<Integer> one,
+                  @StateId("my-state-id") ValueState<Integer> two) {}
+            }.getClass());
+  }
+
+  @Test
+  public void testStateParameterWrongStateType() throws Exception {
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage("has type");
+    thrown.expectMessage("WatermarkHoldState");
+    thrown.expectMessage("but is a reference to");
```

Comment:
Can probably drop the "has type" and "but is a reference to" expectations, without making this test weaker (in fact, it would then be less change-detectory).